src/compiler/checker.ts

```diff
@@ -3995,13 +3995,23 @@ namespace ts {
             if (!(symbol && !isPropertyOrMethodDeclarationSymbol(symbol))) {
                 return undefined;
             }
+            const links = getSymbolLinks(symbol);
+            const cache = (links.accessibleChainCache ||= new Map());
+            // Go from enclosingDeclaration to the first scope we check, so the cache is keyed off the scope and thus shared more
+            const firstRelevantLocation = forEachSymbolTableInScope(enclosingDeclaration, (_, __, node) => node);
+            const key = `${useOnlyExternalAliasing ? 0 : 1}|${firstRelevantLocation && getNodeId(firstRelevantLocation)}|${meaning}`;
+            if (cache.has(key)) {
+                return cache.get(key);

```

Comment:
This cache could probably be better - intelligently sharing results for commonish-`meaning`s and having a hierarchical cache both sound nice in theory, but are a pain to implement performantly (the last time I tried I ruined perf in the general case). This relatively simple cache was sufficient to bring the example runtime down from 20s to 7s on my machine, though is likely to only really effect really degenerate cases, like the one in the linked issue.